lib/rex/proto/kerberos/model/kdc_option_flags.rb

```diff
@@ -0,0 +1,73 @@
+# -*- coding: binary -*-
+
+module Rex
+  module Proto
+    module Kerberos
+      module Model
+        # Represents the KdcOptions KerberosFlags.
+        # https://www.rfc-editor.org/rfc/rfc4120.txt - KDCOptions      ::= KerberosFlags
+        class KdcOptionFlags
+          # @return [Integer] the integer value of the kerberos flags
+          attr_reader :value
+
+          # @param [Integer] value the numerical value of the flags
+          # @raise [ArgumentError] if any of the parameters are of an invalid type
+          def initialize(value)
+            raise ArgumentError, 'Invalid value' unless value.is_a?(Integer)
+
+            @value = value
+          end
+
+          # @param [Array<Integer,Rex::Proto::Kerberos::Model::KdcOptionFlag>] value the numerical value of the flags
+          # @return [Rex::Proto::Kerberos::Model::KdcOptionFlags]
+          def self.from_flags(flags)
+            value = 0
+            flags.each do |flag|
+              value |= 1 << (31 - flag)
+            end
+
+            new(value)
+          end
+
+          def to_i
+            @value
+          end
+
+          # @param [Integer,Rex::Proto::Kerberos::Model::KdcOptionFlag] flag the numerical value of the flag to test for.
+          # @return [Boolean] whether the flag is present within the current KdcOptionFlags
+          def include?(flag)
+            ((value >> (31 - flag)) & 1) == 1
+          end
+
+          # @return [Boolean] whether the flag is present within the current KdcOptionFlags
+          def enabled_flag_names
+            sorted_flag_names = KdcOptionFlag.constants.sort_by { |name| KdcOptionFlag.const_get(name) }
+            enabled_flag_names = sorted_flag_names.select { |flag| include?(KdcOptionFlag.const_get(flag)) }
+
+            enabled_flag_names
+          end
+
+          # Override the equality test for KdcOptionFlags. Equality is
+          # always tested against the #value of the KdcOptionFlags.
+          #
+          # @param [Object] other_object the object to test equality against
+          # @raise [ArgumentError] if the other object is not either another KdcOptionFlags or a Integer
+          # @return [Boolean] whether the equality test passed
+          def ==(other)
+            if other.is_a? self.class
+              value == other.value
+            elsif other.is_a? Integer
+              value == other
+            elsif other.nil?
+              false
+            else
+              raise ArgumentError, "Cannot compare a #{self.class} to a #{other.class}"
+            end
+          end
+
+          alias === ==
+        end
```

Comment:
Why won't you use `BinData` to store [bit fields](https://github.com/dmendel/bindata/wiki/PrimitiveTypes#bit-based-integers) like we usually [do](https://github.com/rapid7/ruby_smb/blob/ef2d2faa0894da614e2aa8bbaba841786989a69c/lib/ruby_smb/smb2/bit_field/directory_access_mask.rb)? I think it would be easier to read and to maintain:
```ruby
class KdcOptionFlags < BinData::Record
  endian :little # or :big
  bit1 :renewable
  bit1 :invalid
  bit1 :allow_post_date
  bit1 :proxy
  bit1 :proxiable
  bit1 :forwarded
  bit1 :forwardable
  bit1 :reserved
  ...
end
```
```
[1] pry(main)> struct = KdcOptionFlags.new(forwardable: 1, forwarded: 1)
=> {:renewable=>0, :invalid=>0, :allow_post_date=>0, :proxy=>0, :proxiable=>0, :forwarded=>1, :forwardable=>1, :reserved=>0}
[2] pry(main)> struct.to_binary_s
=> "\x06"
[3] pry(main)> struct.to_hex.to_i(16)
=> 6
[4] pry(main)> struct.field_names.select {|field| struct[field] == 1 }
=> [:forwarded, :forwardable]
```

Reply:
I know it is off-topics, and probably too late, but I think using `BinData` would be a good idea for all the Kerberos structures in `Rex::Proto::Kerberos::Model`.

Reply:
Most of the models for Kerberos won't require bindata, I'll punt on this for now - and I'll see if I what models we can swap out to use bindata before the feature branch is merged into master

Reply:
+1 under the theory that we all think about data operations on ill-designed protocol elements a little bit differently, so having common structural and implementation semantics (sue me, i like Rust) across objects should help maintainers and newcomers alike.